### DISCUSSION TITLE
my search is broken

### DISCUSSION THREAD

**Discussion by samuelrac** [joshhanley/livewire-autocomplete#8] [Help] [answered]
I did the implementation of autocomplete but my searches are broken, as shown in the image below, I don't know how to solve it, could you help me?
![image](https://user-images.githubusercontent.com/9256333/155865911-ef8ea841-3847-4cd9-a007-d0d72c305c1a.png)

**Comment by joshhanley** [comment, marked as answer]
@samuelrac hard to say without seeing your code, but it looks like you are using Eloquent models, so you might not have specified which property to display.

Have a look at my demo here, you can see it is being passed a list of user models, so I specify `'text' => 'name'` to use the `name` property on the `user` model.

https://github.com/joshhanley/livewire-autocomplete-demo/blob/d11370a995e494d377ffee02efc94d8f2507a51d/resources/views/livewire/main.blade.php#L11-L14

Maybe you need to do something like that?

Hope this helps!

**Comment by samuelrac** [comment]
Hi, thank you very much for your help, this is exactly what I was missing, I am pulling information from a MongoDB database with Eloquent.

However, now I need another help, I'm trying to perform an action when the item is selected in autocomplete, but I'm not getting it.

My component configuration is:

![image](https://user-images.githubusercontent.com/9256333/156064990-80d552e9-fbe0-497f-86ae-1d8df52daeeb.png)

In this case, I need to retrieve the ID of the item that was selected to perform a save action in REDIS, but I am not able to retrieve the ID of the item.



````
<?php

namespace App\Http\Livewire;

use Illuminate\Support\Facades\Redis;

use Livewire\Component;
use App\Models\{
    Client,
    Product
};

class OrderCreate extends Component
{
    public $nomeCliente;
    public $idCliente;

    // Autocomplete VARS
    public $nomeProduto;
    public $idProduto;
    public $products;

    // Order VARS
    public $orderId;
    protected $queryString = ['orderId'];

    // Events VARS
    protected $listeners = ['clientSelectedEvent'];

    // Evento disparado pelo modal de busca de cliente
    public function clientSelectedEvent($id)
    {
        try {
            // Recupera o cliente pelo ID
            $cliente = Client::find($id);
            $this->dispatchBrowserEvent('clientSelectedEvent', [
                'idCliente' => $id,
                'nomeCliente' => $cliente->nomeCliente,
            ]);
        } catch (\Exception $e) {
        }
    }

    public function mount()
    {
        $this->getProduct();
    }

    public function updatedIdProduto()
    {
        $this->nomeCliente = $this->idProduto;
    }

    public function updatedNomeProduto()
    {
        $this->getProduct();
    }

    public function getProduct()
    {
        $this->products = Product::query()
        ->when($this->nomeProduto, function($query, $term) {
            return $query->where('nomeProduto', 'like', '%'. $term .'%');
        })->get();

    }

    public function render()
    {
        //Redis::set('nomeCliente', $this->nomeCliente);
        return view('livewire.order-create');
    }

}
````

**Reply by joshhanley** [reply]
@samuelrac no worries! 🙂

So you should just use an updated hook for your id property.

So in `updatedIdProduto()` you should be able to access the id from `$this->idProduto;` and use that to do what you need with it.

Is that not working?

**Reply by samuelrac** [reply]
Unfortunately it's not working, I created the public function updatedIdProduct() just like to perform the actions of saving in Redis but when I select the item it closes the results listing and nothing happens. I was wondering if it wouldn't be the name of the id field in mongo which is "_id" if that wouldn't be making the system not behave correctly.

![image](https://user-images.githubusercontent.com/9256333/156269860-09714aed-66af-4df0-864b-4a8b09d5d0c1.png)

![image](https://user-images.githubusercontent.com/9256333/156269836-557518e0-d583-4b0f-bad5-64e4d96991f6.png)

I'm following exactly what you put in your tutorial and in your example.

```
<div class="col-md-12">
                    <div class="mb-3">
                        <label class="control-label">Produto {{ $idProduto->nomeProduto ?? null }}</label>
                        <x-lwa::autocomplete
                            name="product-search"
                            wire:model-id="idProduto"
                            wire:model-text="nomeProduto"
                            wire:model-results="products"
                            wire:focus="getProduct"
                            :options="[
                                'id' => '_id',
                                'text' => 'nomeProduto',
                                'allow-new' => 'false',
                            ]"
                        />
                    </div>
                </div>
```

**Reply by joshhanley** [reply]
@samuelrac ah maybe, I didn't realise Mongo has a different id structure. Yeah I think what you have done mapping `'id' => '_id'` seems like the right thing to do. Can you do a dump of one of your products so I can see what its structure looks like so I can confirm that?

Also one other thing I noticed, is you don't have a `$rules` property in your component. Livewire won't send any Eloquent Collection or Eloquent Model data to the front end without it. See the warning box in this section of the Livewire docs https://laravel-livewire.com/docs/2.x/properties#binding-models

So try also adding this to your component, which should allow Alpine to see the `_id` and `nomeProduto` of the products collection.

```php
protected $rules = [
    'products.*._id' => '',
    'products.*.nomeProduto' => '',
];
```

Hope this helps!

**Reply by samuelrac** [reply]
Woooool, it worked perfectly, I was missing the protected $rules, it was very good, one last question please, would it be possible for me to reset the autocomplete field as soon as the item is selected? My idea would be to select the item and fill in a table and add it to redis, and then add another product without having to click on the "X".

Here is my product database structure in MongoDB:
```
{
    "_id" : ObjectId("62128862d936000079006992"),
    "nomeProduto" : "1 produto de teste editado",
    "valorCusto" : 1.5,
    "valorVenda" : 1.9,
    "valorLucro" : 0.4,
    "porcentagemLucro" : 26.67,
    "quantidade" : 100.0,
    "status" : "active",
    "updated_at" : ISODate("2022-02-21T02:38:56.423Z"),
    "created_at" : ISODate("2022-02-20T18:28:50.769Z")
}
```

**Reply by joshhanley** [reply]
@samuelrac nice!

Yeah just set your `$idProduto` and `$nomeProduto` properties to `null` once you have processed the ID and that should clear it out.

**Reply by samuelrac** [reply]
Thank you so much, you have no idea how much you helped me! I've been programming for a while but I'm starting with livewire now and it's been a little tricky for me.

**Reply by joshhanley** [reply]
@samuelrac no worries! Glad it's all sorted for you 🙂
